### PR TITLE
TLS for all paths of traefik_route

### LINKED ISSUE #352: Traefik generates TLS config only for the first defined HTTP router
### Bug Description

In `_update_dynamic_config_route(...`, a TLS config is added to the HTTP routers created via TraefikRoute:
https://github.com/canonical/traefik-k8s-operator/blob/main/src/charm.py#L796

However, if there's more than one router defined in `http` section, it creates the config only for the first entry. It also doesn't create TLS entries for TCP routers.

### To Reproduce

- create a traefik-route relation that defines more than 1 http router (you can use `edge` [tempo-k8s-operator](https://github.com/canonical/tempo-k8s-operator) for that)
- look at generated traefik dynamic config in traefik's container:
- `juju ssh --container traefik traefik/0`

- see that a separate `{ROUTER_NAME}-tls` config is created only for the first http router, pointing at `websecure` entrypoint regardless of what entrypoint the original router points to:

```
root@traefik-0:/etc/traefik# cat /opt/traefik/juju/juju_ingress_traefik-route_49_tempo.yaml 
http:
  routers:
    juju-test-tls-tempo-jaeger-thrift-http:
      entryPoints:
      - jaeger-thrift-http
      rule: ClientIP(`0.0.0.0/0`)
      service: juju-test-tls-tempo-service-jaeger-thrift-http
    juju-test-tls-tempo-jaeger-thrift-http-tls:
      entryPoints:
      - websecure
      rule: ClientIP(`0.0.0.0/0`)
      service: juju-test-tls-tempo-service-jaeger-thrift-http
      tls: {}
    juju-test-tls-tempo-otlp-http:
      entryPoints:
      - otlp-http
      rule: ClientIP(`0.0.0.0/0`)
      service: juju-test-tls-tempo-service-otlp-http
    juju-test-tls-tempo-tempo-http:
      entryPoints:
      - tempo-http
      rule: ClientIP(`0.0.0.0/0`)
      service: juju-test-tls-tempo-service-tempo-http
    juju-test-tls-tempo-zipkin:
      entryPoints:
      - zipkin
      rule: ClientIP(`0.0.0.0/0`)
      service: juju-test-tls-tempo-service-zipkin
  services:
    juju-test-tls-tempo-service-jaeger-thrift-http:
      loadBalancer:
        servers:
        - url: http://tempo-0.tempo-endpoints.test-tls.svc.cluster.local:14268
    juju-test-tls-tempo-service-otlp-http:
      loadBalancer:
        servers:
        - url: http://tempo-0.tempo-endpoints.test-tls.svc.cluster.local:4318
    juju-test-tls-tempo-service-tempo-http:
      loadBalancer:
        servers:
        - url: http://tempo-0.tempo-endpoints.test-tls.svc.cluster.local:3200
    juju-test-tls-tempo-service-zipkin:
      loadBalancer:
        servers:
        - url: http://tempo-0.tempo-endpoints.test-tls.svc.cluster.local:9411
tcp:
  routers:
    juju-test-tls-tempo-otlp-grpc:
      entryPoints:
      - otlp-grpc
      rule: ClientIP(`0.0.0.0/0`)
      service: juju-test-tls-tempo-service-otlp-grpc
  services:
    juju-test-tls-tempo-service-otlp-grpc:
      loadBalancer:
        servers:
        - address: tempo-0.tempo-endpoints.test-tls.svc.cluster.local:4317
```

### Environment

`latest/edge` traefik

### Relevant log output

```shell
code already shows the problem
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mmkay**
## Issue
Fixes #352.

## Solution
Apply the same logic to all routers, including TLS routers.

## Testing Instructions
Bundle traefik together with `edge` Tempo and peek at a generated route file inside Traefik container.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -796,23 +796,43 @@ def _provide_routed_ingress(self, relation: Relation):
...
-                )
-
+        if "tcp" in config.keys():
+            route_config = config["tcp"].get("routers", {})
+            # here we modify entries instead, adding a passthrough section, so we don't copy
+            for router_name in route_config.keys():
+                router_config = route_config.get(router_name, {})
+                route_rule = router_config.get("rule", "")
+                service_name = router_config.get("service", "")
+                if not all([router_name, route_rule, service_name]):
```
[the remark this answers is not in the document]
Comment (mmkay) [reply src/charm.py:831]: Let's say it's a defensive check for whether this router config seems not to be an empty dict where we wouldn't want to add the passthrough section

[the remark this answers is not in the document]
Comment (mmkay) [reply src/charm.py:838]: Now we do, after I changed the code to not use passthrough anymore.

File: src/charm.py [hunk 2]
```diff
@@ -796,23 +796,58 @@ def _provide_routed_ingress(self, relation: Relation):
...
+            route_config = config["tcp"].get("routers", {})
+            # we want to generate and add a new router with TLS config for each routed path.
+            # as we mutate the dict, we need to work on a copy
+            for router_name in route_config.copy().keys():
+                route_rule = route_config.get(router_name, {}).get("rule", "")
+                service_name = route_config.get(router_name, {}).get("service", "")
+                entrypoints = route_config.get(router_name, {}).get("entryPoints", [])
+                if len(entrypoints) > 0:
+                    # for grpc, all entrypoints are custom
+                    entrypoint = entrypoints[0]
```
Comment (PietroPasotti) [inline src/charm.py:834]: what are we doing with the other entrypoints? are we only adding tls to the first one?

**Review by PietroPasotti — APPROVED** [review]
nice and self-contained. 
Would like to see a test for this but let's do that in a separate PR, it will be easier when we have tempo tls ready too

File: src/charm.py (hunk 2 above)
Comment (mmkay) [reply src/charm.py:834]: In reality we always have one entrypoint in our routers. Might be worth adding better handling later for more custom scenarios.

**Pull request merged**